### DISCUSSION TITLE
Climate set temperature in Home assistant

### DISCUSSION THREAD

**Discussion by heijjan** [visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA#408] [General]
Hi,

Everything has been working quite nicely for two weeks, but one thing is weird - When I hit the Climate: set temperature function in Home Assistant, the temperature will go up and down multiple times before settling down to the desired target causing the heater to go on and off multiple times also. Is this something that should happen or could this be fixed?

Here's how it went last night while the automation asked for a target temp of 30C:

![layz](https://user-images.githubusercontent.com/111536037/201865871-31bd2d07-4560-4113-8085-9511fcbccb9c.png)

**Comment by visualapproach** [comment]
Hi, what fw version?

**Reply by heijjan** [reply]
Hi, this is the development release from a week back or so.

MQTT: CONNECTED
Firmware version: 2022-11-07
Model: MIAMI2021
RSSI: -66

**Reply by visualapproach** [reply]
Thx, I have to check. It shouldn't do that. Will look at it in a few days

**Reply by visualapproach** [reply]
@heijjan I cannot reproduce, but I have another model. If you are sure the fault is not on the HA side*, you could try this: Find this code in "lib/BWC/BWC_common.cpp"
```
            int pushtime = 500;
            int releasetime = 300;
```
It's at lines 536-537 in latest dev version. Inside function `_handleCommandQ`
Change the releasetime to 1000. It will take more time to get to new setpoint, but it shouldn't overshoot.

*you can for instance download and run "MQTT explorer", connect it to your broker, and watch the traffic on the /command topic. When setting target temp it should only send one message. (It will send Fahrenheit but it will get converted to Celsius in the ESP if needed)

**Reply by heijjan** [reply]
Thank you. This didn't really help, the temperature is still fluctuating a lot, there must be a hardware issue or sth. 

I get these multiple times before the temp settles down to the requested target temp:

`{
    "CONTENT": "OTHER",
    "MQTT": 0,
    "PressedButton": "1b1b",
    "HASJETS": false,
    "RSSI": -66,
    "IP": "10.1.20.6",
    "SSID": "VHT30B4_IOT",
    "FW": "2022-11-07",
    "MODEL": "MIAMI2021"
}`

And 
`{
    "CONTENT": "STATES",
    "TIME": 1668784916,
    "LCK": 0,
    "PWR": 1,
    "UNT": 1,
    "AIR": 0,
    "GRN": 0,
    "RED": 0,
    "FLT": 1,
    "CH1": 32,
    "CH2": 51,
    "CH3": 48,
    "HJT": 0,
    "BRT": 7,
    "TGT": 30,
    "TMP": 21,
    "VTF": 20.80340767,
    "VTMC": 20.76375389,
    "VTMF": 69.37475586,
    "AMBC": -2,
    "AMBF": 28,
    "AMB": -2,
    "VTM": 20.76375389,
    "TGTC": 30,
    "TMPC": 21,
    "TGTF": 86,
    "TMPF": 70
}`

And

`{
    "CONTENT": "TIMES",
    "TIME": 1668784916,
    "CLTIME": 1668198268,
    "FTIME": 1668294670,
    "UPTIME": 1212039,
    "PUMPTIME": 638152,
    "HEATINGTIME": 187024,
    "AIRTIME": 78,
    "JETTIME": 0,
    "COST": 15.97325611,
    "FINT": 30,
    "CLINT": 14,
    "KWH": 106.4883728,
    "KWHD": 16.66552111,
    "WATT": 42,
    "TTTT": -11318,
    "T2R": "Never",
    "MINCLK": 5
}`

**Reply by visualapproach** [reply]
okay, that's completely normal. Those messages are sent as soon as anything worth mentioning has changed. If we try to narrow things down a bit by not considering your automation - Can you check and describe what's happening when you change target temperature from the web GUI. Does it overshoot? And what happens when you change from the climate control  in HA? It's supposed to tick up or down to the new setpoint in a couple of seconds, and then stop.

**Reply by visualapproach** [reply]
Worth noting- the "dial" should step 1 degree at a time. Not more. Like 27-28-29-30-stop. If it makes 27-29-31-... something is wrong.

**Reply by heijjan** [reply]
The functionality seems to be more or less the same while adjusting the temp from HA's climate control slider vs. setting the target temp via automation. There doesn't seemt to be any logic how the adjustment goes, I just set the target to 40C and the temp go to the desired 40C with increments of 2-3 from 30C -> 32 -> 35 -> 37 -> 40. And from 40C to 22C it went like 40 -> 35 -> 20 -> 26 -> 29 -> etc and ended to 22C after a minute or so. This is the message that keeps repeating with different "TGTC" value:

`{
    "CONTENT": "STATES",
    "TIME": 1668799250,
    "LCK": 0,
    "PWR": 1,
    "UNT": 1,
    "AIR": 0,
    "GRN": 0,
    "RED": 0,
    "FLT": 1,
    "CH1": 32,
    "CH2": 50,
    "CH3": 54,
    "HJT": 0,
    "BRT": 7,
    "TGT": 26,
    "TMP": 20,
    "VTF": 20.98999977,
    "VTMC": 20.98802376,
    "VTMF": 69.77844238,
    "AMBC": -3,
    "AMBF": 27,
    "AMB": -3,
    "VTM": 20.98802376,
    "TGTC": 26,
    "TMPC": 20,
    "TGTF": 79,
    "TMPF": 68
}`

**Reply by visualapproach** [reply]
And from the web gui ?

**Reply by heijjan** [reply]
More or less the same, target to 30C: 20C -> 24C -> 27C -> 32C -> 26C -> 27C -> 31C -> 30C.

**Reply by visualapproach** [reply]
Ok, how does the display on the pump behave? Is every degree shown? How long between changes?

**Reply by heijjan** [reply]
The display of the pump shows all the numbers going up and down every second or so. Sometimes the temp reading goes black for a while and comes back up and as soon as the pump has reached the target temperature, the pump display starts showing the current water temp. 

The same goes other way around, if I try to adjust the temp from the pump display, the temp may go up for 10 degrees at once and come back 5 degrees after that etc. I resoldered a new pair of wires today to the PCB to make sure that there aren't any broken wires in the circuit.

**Reply by visualapproach** [reply]
That's the problem. The way the ESP knows, well, anything, is by reading the display. Actually by reading what the CIO sends to the display and then retransmit it to the display. But if it shows wrong or no info it cannot know what the new target temperature is. There is a few ways to check what's wrong.
1. hwtestinfo.html will check every output/input on the device. You should see 0 errors preferably.
2. Is the display showing text messages correctly? Like the fw version at startup. If it does then esp-dsp is likely fine. And the esp-cio connection may be bad.
3. Monitor the 'pressed button' in HA while pressing different buttons. If they are wrong then there is a problem reading the dsp buttons. Else there is probably com problems with the cio.

I recommend you do all checks above. And (I'm on the cellphone now and can't remember) post a pic of your device if you haven't.
I think you may be better off using pinout according to pcb_v2b (IIRC). See #330

**Reply by heijjan** [reply]
Alright, I need to check these when it is again warmer than -5C or the electricity is cheap so that there is room for failures. 

Do you mean that I could change the PVC version to 2 in the code even though I’m running a V1 Board or do you mean that I should change the wiring on the board?

**Reply by visualapproach** [reply]
You don't need to change the physical PCB. The #define PCB_Vx in the code is just a predefined set of pin outs that tells the ESP which pins do what. V1 and V2 matches the respective PCB layout. V2B is using pins which are most likely to work with the new pumps. You can use it with any PCB but you have to solder the wires to the corresponding pins. In your case the lower left in this picture.
![image](https://user-images.githubusercontent.com/22987250/202862244-6cddf2b9-c310-4432-8f0c-08edae54a830.png)

**Reply by heijjan** [reply]
Ok, thank you. Just a check - I have soldered the wires as per the instructions, could you clarify which wires should be soldered elsewhere? 

![solders](https://user-images.githubusercontent.com/111536037/202864325-2c1c6e5c-6c2b-4bbc-88ff-333ac27fd66f.png)

Did I understood correctly that the needed changes are:

H1 (the male connector)
- Wire 1 (white): No change
- Wire 2 (blue): No change
- Wire 3 (red): Change from 3. hole to 5. hole
- Wire 4 (green): No change
- Wire 5 (yellow): Change from 5. hole to H2 connector 3. hole
- Wire 6 (black): Wire not connected -> To be removed

H2 (the female connector)
- Wire 1 (white): No change
- Wire 2 (blue): No change
- Wire 3 (red): Change from 3. hole to 6. hole
- Wire 4 (green): No change
- Wire 5 (yellow): No change
- Wire 6 (black): Change from 6. hole to H1 connector 3. hole

And after resoldering the wires, I'll change the layout in model.h from #define PCB_V1 -> #define PCB_V2 and reupload the firmware?

**Reply by visualapproach** [reply]
> H1 (the male connector)

> Wire 6 (black): Wire not connected -> To be removed

DOES NOT MATTER


> And after resoldering the wires, I'll change the layout in model.h from #define PCB_V1 -> #define PCB_V2 and reupload the firmware?

No, #define PCB_V2B

The rest seems right

**Reply by heijjan** [reply]
Ok, thank you. I did the modifications and now the display of the pump doesn't stop flashing, controlling the pump over HA works though. Same problem with the target temp adjustment, but functionalities work.

This is how it looks like:

![WhatsApp Image 2022-11-19 at 21 34 14 (1)](https://user-images.githubusercontent.com/111536037/202868344-ca07e9cb-3345-4c33-b0d1-c903ffcf2d06.jpeg)
![WhatsApp Image 2022-11-19 at 21 34 14](https://user-images.githubusercontent.com/111536037/202868347-b8197934-0ea0-4f11-9d87-d5ac7b0209e3.jpeg)
![WhatsApp Image 2022-11-19 at 21 34 13](https://user-images.githubusercontent.com/111536037/202868349-512fe0dc-33a2-44e0-919c-56342ebc2ac0.jpeg)

**Reply by visualapproach** [reply]
You can try without the LLC. Solder directly to the esp pins

**Reply by heijjan** [reply]
Alright, so I could solder the wires directly to the ESP without the PCB at all? Could you point out which wires should go to where? Or should I jump wire somehow the LLC leg points?

**Reply by visualapproach** [reply]
Just solder them to the esp pins with the same label. Wire at PCB D1 moves to ESP D1 and so on. Since you soldered the LLC directly to the board it's a PITA to remove without proper equipment and it's better to do this way. You may leave 5V and GND wires as is.

**Reply by heijjan** [reply]
Thanks! I changed the wiring directly to the NodeMCU pins and with my other NodeMCU, it starts fine with the melody and the display shows the current temperature, lights are not flashing or anything, but the button presses do nothing and adjusting from web gui or HA doesn't work. From MQTT log I can see that it is pressing all the buttons all the time. The display of the pump doesn't show anything else but the current water temp.
![Untitled](https://user-images.githubusercontent.com/111536037/202891667-10d8911a-e43d-4c98-b447-30a3642f5c81.png)

I have uploaded the firmware with PCB_V2B layout to that board.

The other NodeMCU, which worked well previously seems to be blown. When I connect it to the pump, the pump display starts to make a rattling noise in accordance with the WIFI indicator led's blinking sequence in the NodeMCU board. 

In case I ordered just another new NodeMCU, should it work without the PCB so that I solder the ground wires to the G legs of the NodeMCU and 5V wires to the VIN leg of the NodeMCU + other wires to the corresponding legs? And then the firmware with PCB_V2B layout, or does it even matter?

**Reply by visualapproach** [reply]
> Thanks! I changed the wiring directly to the NodeMCU pins and with my other NodeMCU, it starts fine with the melody and the display shows the current temperature, lights are not flashing or anything, but the button presses do nothing and adjusting from web gui or HA doesn't work. From MQTT log I can see that it is pressing all the buttons all the time. The display of the pump doesn't show anything else but the current water temp.
> 
> ![Untitled](https://user-images.githubusercontent.com/111536037/202891667-10d8911a-e43d-4c98-b447-30a3642f5c81.png)
> 
> 
> 
> I have uploaded the firmware with PCB_V2B layout to that board.
> 
> 
> 
> The other NodeMCU, which worked well previously seems to be blown. When I connect it to the pump, the pump display starts to make a rattling noise in accordance with the WIFI indicator led's blinking sequence in the NodeMCU board. 
> 
> 
> 
> In case I ordered just another new NodeMCU, should it work without the PCB so that I solder the ground wires to the G legs of the NodeMCU and 5V wires to the VIN leg of the NodeMCU + other wires to the corresponding legs? And then the firmware with PCB_V2B layout, or does it even matter?
> 
> 

You don't need a physical PCB if soldering directly to the nodeMCU. Yes ground to G and 5V to Vin. And the other wires to pins according to PCB_V2B. You need to define the pcb_v2B and model miami2021. 

Those wires can be of poor quality. Check or replace them. Use flux when soldering.
I think the different behavior between the two nodeMCUs points to something hardware related, given the software is the same.

**Reply by heijjan** [reply]
Right. I'll order new cables just to be on a safer side together with new NodeMCU.

**Reply by heijjan** [reply]
Alright, I have now 3 brand new NodeMCU's and 6 pairs of wires. I have soldered one together and it is giving me the same result as the one, which makes the display rattle with the sequence of the wifi led and nothing else happens with the display. I managed to get one 5V voltage supply connected to the NodeMCU board and the result is:

`DSP input pin 2 500 errors of 1000
DSP input pin 3 0 errors of 1000
DSP input pin 4 500 errors of 1000
CIO input pin 2 500 errors of 1000
CIO input pin 3 0 errors of 1000
CIO input pin 4 500 errors of 1000`

The I put another, older one together with different batch wires and the same end result, same errors. 

For me it seems rather impossible that different batch of wires, different batch of NodeMCU's set together at different times give the exact same errors. Is it possible that there is something wrong with the code?

The wires are now soldered directly to the NodeMCU's and the pinout is set to V2B in the code. The other board is using the latest development firmware and the other board latest master release. I just want to make sure that this isn't a software issue before starting to build another one.

**Reply by visualapproach** [reply]
You just reminded me that I should update the hardware test report. It's a bit confusing now since pin2 is the 3rd pin. It happened to start counting from zero. And 500/1000 meaning NO CONNECTION AT ALL. It doesn't mean 50% good, which you may be led to believe.
Anyhow, I strongly suspect you have soldered according to PCB_V1 and defined PCB_V2B or vice versa. Both definitions use the same pins for the CLK which is the only one without errors in your hwtest report. Remember to save all edited files before compiling/uploading.

Make sure PCB_V2B is defined (and no other PCB defined)
The first and second wire is 5V and GND. The third wire from the CIO (pump) connected to D1. 4th wire to D2 and 5th wire to D5.  
From the display connect 5V, GND and then third wire to D6, 4th to D4, 5th to D3 and last one to D7.
If you soldered your wires this way without shorts or open circuits you should see 0/1000 on all pins after the hwtest.

It will only work when you see all 0/1000 in the test. If you don't there is something wrong with your wiring.

**Reply by heijjan** [reply]
Thanks! I had soldered the pins in wrong order, I didn't realize that the pin numbering changed. I ran the test without any errors and connected the module to the pump. The display played the melody, the display showed all messages and shows current temperature, but the buttons don't do anything nor does the remote control from the web ui or from HA. The pump stays idle.

Here's what I can see from MQTT log

Trying to start the pump from HA:

`Viesti 13 vastaanotettu aiheessa layzspa/command 21.02:
{CMD:4,VALUE:true,XTIME:0,INTERVAL:0}`

Trying to start the heat regulation from HA:
`{CMD:3,VALUE:true,XTIME:0,INTERVAL:0}`

Nothing happens.

**Reply by visualapproach** [reply]
It seems that the communication with the CIO is problematic. It got worse after getting rid of the LLC. The easy fixes seem to not work in this case. My best guess would be to buy one of those red LLCs with transistors on it instead of the IC. You'd need the new PCB_V2 though. And it is still no guarantee it will work.

**Reply by heijjan** [reply]
Alright, thank you. I'll order the red LLC, some pin headers and V2 PCBs. Hopefully I'll be testing this out again in couple of weeks.

**Reply by heijjan** [reply]
You can't believe what just happenned. I took another look at the original board and its configuration, I replaced the cables, rechecked the solders, erased the flash and replaced it with the latest dev release with PCB_V1 configuration, ran the hwtest with 0 errors and installed it to the pump - It works. Even the temp control from HA without any hazzle. This was the board that I thought I've blown during testing and never retested with that. 

Now I have a working configuration with PCB_V1 and new parts for PCB_V2 coming in following weeks. Let's see if I have energy to test with those too!

Thank you again about your great support.

**Reply by Jay-Ti** [reply]
Yep... I got the feeling theres some black voodoo magic either in the esps or thr code.
Know your story too well, although i have come around and the once working unmodified code now either crashes or sees the dsp as dead 😂
Even soldered a new board, checked continuity, only thing left is use a signal generator and Oszilloscope to check the cables.

**Reply by visualapproach** [reply]
@heijjan no problemo. Glad it works now.

@Jay-Ti Black magic LoL. No clue what's going on but maybe the signals just barely works and when someone farts nearby that's enough to tip it over.
